### PR TITLE
Migrate 3 utils and 5 runtime file to TypeScript

### PR DESCRIPTION
**Pull request by ChrisAntaki**
- Migrate files to TypeScript
- Improve typing in `src/model/doc`
- Update rules

### CODE REVIEW

File: build-system/tasks/check-rules.js [hunk 1]
```diff
@@ -56,10 +56,6 @@ const forbiddenTerms = {
     allowlist: ['src/main.js', 'src/components/activities.js'],
     checkInTestFolder: true,
   },
-  '(?:var|let|const) +IS_DEV +=': {
-    message: 'IS_DEV local var only allowed in mode.js',
-    allowlist: ['src/mode.js'],
```
Comment (ChrisAntaki) [inline build-system/tasks/check-rules.js:61]: Note: This file no longer exists

File: build-system/tasks/check-rules.js [hunk 2]
```diff
@@ -157,20 +149,12 @@ const forbiddenTermsSrcInclusive = {
...
-    allowlist: ['src/utils/date-utils.js'],
-  },
   '\\<\\<\\<\\<\\<\\<': {
     message: 'Unresolved merge conflict.',
   },
   '\\>\\>\\>\\>\\>\\>': {
     message: 'Unresolved merge conflict.',
   },
-  '\\.trim(Left|Right)\\(\\)': {
-    message: 'Unsupported on IE; use trim() or a helper instead.',
```
Comment (ChrisAntaki) [inline build-system/tasks/check-rules.js:171]: Note: Swgjs no longer supports IE

File: build-system/tasks/check-rules.js [hunk 3]
```diff
@@ -159,18 +151,13 @@ const forbiddenTermsSrcInclusive = {
   },
   '\\.getTime\\(\\)': {
     message: 'Unless you do weird date math (allowlist), use Date.now().',
-    allowlist: ['src/utils/date-utils.js'],
```
Comment (ChrisAntaki) [inline build-system/tasks/check-rules.js:162]: Note: The file wasn't using `.getTime()`

File: src/runtime/experiments.ts [hunk 4]
```diff
@@ -194,20 +182,17 @@ function parseSetExperiment(win, experimentMap, spec) {
...
       // Ignore: experiment parsing cannot block runtime.
       on = false;
-      ErrorUtils.throwAsync(e);
+      ErrorUtils.throwAsync(e as Error);
     }
   } else {
     on = false;
   }
 
+  experimentMap ||= {};
```
Comment (fikus) [inline src/runtime/experiments.ts:191]: Can you change the type of experimentMap so it is not nullable, instead?

**Review by fikus — APPROVED** [review]

File: src/runtime/experiments.ts (hunk 4 above)
Comment (ChrisAntaki) [reply src/runtime/experiments.ts]: Updated ✅

**Pull request merged**